File: code/ex134.java
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int n = gas.length;
        
        int idx = -1, sum = 0, min = Integer.MAX_VALUE;
        for (int i = 0; i < n; i++) {
            sum = sum + gas[i] - cost[i];
            if (sum < min) {
                idx = i;
                min = sum;
            }
        }
        if (sum < 0)    return -1;    // Cannot return whichever the start point is
        
        return idx + 1 < n ? idx + 1 : 0;
    }
}

File: code/ex134.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int n = gas.size();
        
        int idx = -1, MIN = INT_MAX, prefix = 0;
        for (int i = 0; i < n; i++) {    // Find the idx where prefix sum is minimum
            prefix += gas[i] - cost[i];
            if (prefix < MIN) {
                MIN = prefix;
                idx = i;
            }
        }
        if (prefix < 0)    return -1;
        
        return idx + 1 < n ? idx + 1 : 0;
    }
};